### PR TITLE
Refactor loadaer in ndrop.F90

### PR DESCRIPTION
**Pull request by brigaudet**
Refactor ndrop.F90/loadaer subroutine.  Includes:  1)  retrieve mixing ratios and number concentrations by directly referencing state q vector instead of rad_cnst procedures; 2)  remove unused branches for phase == 2 (cloud only); 3) condense number of loops over columns.

### CODE REVIEW

File: components/eam/src/physics/cam/ndrop.F90 [hunk 1]
```diff
@@ -1521,106 +1524,128 @@ end subroutine ccncalc
...
+             vol = max(raer(icol) + qqcw(icol), 0._r8)/specdens_val
+         else if (phase == 1) then
+             vol = max(raer(icol), 0._r8)/specdens_val
+         endif
+
+         vaerosol(icol) = vaerosol(icol) + vol
+         hygro(icol)    = hygro(icol) + vol*spechygro_val
+
+         if( last_pass ) then
+             if (vaerosol(icol) > 1.0e-30_r8) then   ! +++xl add 8/2/2007
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/ndrop.F90]: this comment may be removed

File: components/eam/src/physics/cam/ndrop.F90 [hunk 2]
```diff
@@ -1521,106 +1524,128 @@ end subroutine ccncalc
...
+       mindex = numptrcw_amode(imode)
+       qqcw = state%q(:,kk,mindex) 
+   endif
+
+   do icol = istart, istop
+       if (phase == 3) then
+          naerosol(icol) = (raer(icol) + qqcw(icol))*cs(icol,kk)
+       else
+          naerosol(icol) = raer(icol)*cs(icol,kk)
+       end if
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/ndrop.F90]: `end if`  --> `endif`

File: components/eam/src/physics/cam/ndrop.F90 [hunk 3]
```diff
@@ -1521,106 +1524,128 @@ end subroutine ccncalc
...
    ! adjust number so that dgnumlo < dgnum < dgnumhi
-   do i = istart, istop
-      naerosol(i) = max(naerosol(i), vaerosol(i)*voltonumbhi_amode(m))
-      naerosol(i) = min(naerosol(i), vaerosol(i)*voltonumblo_amode(m))
+      naerosol(icol) = max(naerosol(icol), vaerosol(icol)*voltonumbhi_amode(imode))
+      naerosol(icol) = min(naerosol(icol), vaerosol(icol)*voltonumblo_amode(imode))
    end do
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/ndrop.F90]: `end do` --> `enddo`

File: components/eam/src/physics/cam/ndrop.F90 [hunk 4]
```diff
@@ -1521,106 +1524,128 @@ end subroutine ccncalc
...
+   do lspec = 1, nspec_amode(imode)
+      first_pass = .false.
+      last_pass = .false.
 
-   do l = 1, nspec_amode(m)
+      if( lspec == 1 ) then
+          first_pass = .true.
+      elseif( lspec == nspec_amode(imode) ) then
+          last_pass = .true.
+      end if
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/ndrop.F90]: `end if` --> `endif`

**Review by tangshuaiqi — APPROVED** [review]
only have a few minor comments.

**Review by jeff-cohere — APPROVED** [review]

**Review by mingxuanwupnnl — APPROVED** [review]

File: components/eam/src/physics/cam/ndrop.F90 [hunk 5]
```diff
@@ -1521,109 +1524,187 @@ end subroutine ccncalc
...
+     qcldbrn_num(:) = fldcw(:,kk)
+  endif
 
-   ! return aerosol number, volume concentrations, and bulk hygroscopicity
+  !Note for C++ port: Code to be ported starts below
 
-   ! input arguments
-   type(physics_state), target, intent(in) :: state
-   type(physics_buffer_desc),   pointer    :: pbuf(:)
+  !Currenly supports only phase 1 (interstitial) and 3 (interstitial+cldbrn)
```
Comment (singhbalwinder) [inline components/eam/src/physics/cam/ndrop.F90]: This should be the first thing to check, so it should be moved to the top of this subroutine.

**Review by singhbalwinder — APPROVED** [review]
@brigaudet : I have suggested only one edit. If everything looks good, you can run the long tests to see if they all PASS. Thanks!

File: components/eam/src/physics/cam/ndrop.F90 (hunk 5 above)
Comment (brigaudet) [reply components/eam/src/physics/cam/ndrop.F90]: OK, this has been done in the latest commit.  Looks like all the longest tests pass:

comp_refac_loadaer_ex_mam4_org_v2_baselines: 5 tests
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel

**Pull request merged**